go/arrow/array/list.go

```diff
@@ -1411,217 +1410,56 @@ func (b *baseListViewBuilder) UnmarshalJSON(data []byte) error {
 			minOffset = offset
 		}
 	}
 	return minOffset
 }
 
-// Find the maximum offset+size in a LARGE_LIST_VIEW array.
+// Find the maximum offset+size in a LIST_VIEW/LARGE_LIST_VIEW array.
 //
 // Pre-conditions:
 //
-//	input.DataType() is LargeListViewType
-//	input.Len() > 0 && input.NullN() != input.Len()
-func maxLargeListViewOffset64(input arrow.ArrayData) int64 {
+//	input.DataType() is ListViewType if Offset=int32 or LargeListViewType if Offset=int64
+//	input.Len() > 0
+func maxListViewEnd[Offset int32 | int64](input arrow.ArrayData) Offset {
 	inputOffset := input.Offset()
-	var bitmap []byte
-	if input.Buffers()[0] != nil {
-		bitmap = input.Buffers()[0].Bytes()
-	}
-	offsets := arrow.Int64Traits.CastFromBytes(input.Buffers()[1].Bytes())[inputOffset:]
-	sizes := arrow.Int64Traits.CastFromBytes(input.Buffers()[2].Bytes())[inputOffset:]
+	offsets := arrow.GetData[Offset](input.Buffers()[1].Bytes())[inputOffset:]
+	sizes := arrow.GetData[Offset](input.Buffers()[2].Bytes())[inputOffset:]
 
-	isNull := func(i int) bool {
-		return bitmap != nil && bitutil.BitIsNotSet(bitmap, inputOffset+i)
-	}
+	maxLegalOffset := Offset(input.Children()[0].Len())
 
-	// It's very likely that the first non-null non-empty list-view starts at
-	// offset zero, so we check that first and potentially early-return a 0.
-	i := input.Len() - 1 // safe because input.Len() > 0
-	for i != 0 && (isNull(i) || sizes[i] == 0) {
-		i -= 1
-	}
-	offset := offsets[i]
-	size := sizes[i]
-	if i == 0 {
-		if isNull(i) || sizes[i] == 0 {
-			return 0
-		} else {
-			return offset + size
-		}
-	}
+	i := input.Len() - 1
+	maxEnd := offsets[i] + sizes[i] // safe because input.Len() > 0
 
-	if offset > math.MaxInt64-size {
-		// Early-exit: 64-bit overflow detected. This is not possible on a
-		// valid list-view, but we return the maximum possible value to
-		// avoid undefined behavior.
-		return math.MaxInt64
-	}
-	values := input.Children()[0]
-	maxEnd := offsets[i] + sizes[i]
-	if maxEnd == int64(values.Len()) {
-		// Early-exit: maximum possible view-end found already.
-		return maxEnd
-	}
-
-	// Slow path: scan the buffers entirely.
-	for ; i >= 0; i -= 1 {
-		offset := offsets[i]
-		size := sizes[i]
-		if size > 0 && !isNull(i) {
-			if offset+size > maxEnd {
-				if offset > math.MaxInt64-size {
-					// 64-bit overflow detected. This is not possible on a valid list-view,
-					// but we saturate maxEnd to the maximum possible value to avoid
-					// undefined behavior.
-					return math.MaxInt64
-				}
-				maxEnd = offset + size
-				if maxEnd == int64(values.Len()) {
-					return maxEnd
-				}
-			}
+	for i -= 1; i >= 0; i -= 1 {
+		if maxEnd == maxLegalOffset {
+			// Fast path: the maximum offset+size is frequently exactly the end of the child array,
+			// and frequently a view which has this offset+size will be near the end of the array.
+			return maxEnd
+		}
+		if end := offsets[i] + sizes[i]; end > maxEnd {
+			maxEnd = end
```

Comment:
You can take a look at my final version of this function in the C++ implementation: 

https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/list_util.cc#L91